### ISSUE TITLE
Individualized Bayesian Knowledge Tracing Models

### ISSUE THREAD

**Issue by GokulNC** [myudelson/hmm-scalable#2]
Hello @myudelson !
Thanks a lot for this work, it's really helpful.
Just wanted to know this.

Do you support student-specific parameters in this repo as described in your paper:
[Individualized Bayesian Knowledge Tracing Models
](https://link.springer.com/chapter/10.1007/978-3-642-39112-5_18)

**Comment by myudelson** [comment]
Hi, Gokul,

Yes, those solvers are supported. The solver-algorithm specification -s is
like structure.algorithm, where algorithms are 1 Baum-Welsh, 2 - gradient
descent, 3.1,3.2,3.3,3.4 – variants of conjugate gradient descent, 4 -
lagrangian.

The structure described in the readme is 1 – all by skill.
You would need different structures:
4 - individualized pLo
5 - individualized pLo and pT
6 - individualized pT
7 - individualized pLo, pT, pS, and pG
Baum-Welsh doesn't work with these.

In the Individualized BKT paper I've used -s 4.2 and -s 5.2 settings.

Hope it helps,
Best,
Michael


--
Michael V. Yudelson
Sr. Software Engineer
ACT, Inc.

**Comment by GokulNC** [comment]
Thank you so much.

I was able to use the model variant: "_Individualized BKT with student-specific p(T)_"
as concluded as best setting in the paper, using the argument `-s 6.2` ([STRUCTURE_Agk](https://github.com/myudelson/hmm-scalable/blob/d1da8ec/utils.h#L116))

**Comment by myudelson** [comment]
Gokul,

Setting -s 6.2 (STRUCTURE_Agk) individualizes pT only (A-matrix for state
transitions).

Best,
Michael


--
Michael V. Yudelson
Sr. Software Engineer
ACT, Inc.

**Comment by GokulNC** [comment]
Thanks @myudelson .
The reason I used `STRUCTURE_Agk` was because, this is what your paper quoted:

> The BKT model variants we fit were:
> 1. Standard BKT model,
> 2. Individualized BKT with student-specific p(L0),
> 3. Individualized BKT with student-specific p(T),
> 4. Individualized BKT with student-specific p(L0) and p(T)
> 
> Introduction of student specific speed of learning (p(T)) in model 3 results in a consistent and more pronounced advantage over models 1 and 2. Moreover, the improvement in model accuracy resulting from adding individualized p(L0) on top of individualized p(T) (going from model 3 to model 4) is even smaller than when adding individualized p(L0) to the standard BKT model (going from model 1 to model 2). Given that, **model 3 with individualized p(T) can be considered superior to the standard BKT and other individualized models**.

Also, it would be great if you can reference about this individualized BKT paper in your README and mention the correspondng `-s` arguments

**Comment by myudelson** [comment]
You are absolutely right, I should update the readme to include all the
options.
It's been long pending
Michael

--
Michael V. Yudelson
Sr. Software Engineer
ACT, Inc.

**Comment by GokulNC** [comment]
Hi @myudelson 

I was trying to reproduce your results from the Individualized BKT paper for the `algebra_2008_2009` dataset.

As mentioned in the `README`, I did this to train:
```
./trainhmm -s 6.2 -d ~ -m 1 -p 1 a89_kts_train.txt model.txt predict.txt
```

And tried to evaluate the corresponding test set on the trained model:
```
./predicthmm -d ~ -p 1 a89_kts_test.txt model.txt predict.txt
```

I get the following result on the test:
```
LL= 762941.2809722 ( 161943.3580356), AIC=1539504.561944, BIC=1615379.308948, RMSE=0.737935 (0.506752), Acc=0.000000 (0.000000)
```

Please note that the accuracy is 0.
But if use structure `-s 1.1` and repeat the above procedure, I seem to be getting a 7% accuracy.

Can you please point out if I am missing something?

**Comment by GokulNC** [comment]
Sorry, my bad. It seems like the test file `algebra_2008_2009_test.txt` doesn't have the ground truth, hence all the entries in the processed test file were 2 (NOT_CORRECT). Hence the zero accuracy probably.

But I do not understand why it gave a non-zero accuracy for `-s 1.1`. 

---

Anyway, I ran using a 5-fold cross validation using the following command:
```
./trainhmm -s 6.2 -v 5,g,1 -d ~ -p 1 a89_kts_train.txt model.txt predict.txt
```

And it gives me the following results:
```
5-fold cross-validation: LL=3739584.1295321 (2118847.0982774), AIC=7492980.259064, BIC=7589689.039997, RMSE=0.356673 (0.389590), Acc=0.854771 (0.825732)
```

I hope this corresponds to the average accuracy of the validation set split out from the full set in 1:4 (test:train) ratio.
Please let me know if I am wrong.

I also see the following message being printed, but I believe this is just a warning:
```
Cross-validation is done wrong, if A or B matrices are 'sliced'
```

**Comment by myudelson** [comment]
Hi Gokul,
Did you inspect the prediction file? Are there any anomalies you can spot?
Don't mind the warning about slicing, you are not slicing.
Bear in mind, that "train" data precedes "test" data. In order to properly
predict "test", it should be combined together, otherwise, the prehistory
of student work is lost.
Best,
Michael



--
Michael V. Yudelson
Sr. Software Engineer
ACT, Inc.

**Comment by GokulNC** [comment]
Thanks Mikhail. For each student, I will concatenate their past history (from the train set) before the test data (of each student) as you have pointed out to do evaluation.

But, how will the `predicthmm` to evaluate only the rows corresponding to test set and return only its accuracy?

Thanks!

**Comment by myudelson** [comment]
One way to achieve that is to append test file after to the train file and,
after predicting on the joined file, to detach the test part (lines after
the last line of train file).
Since there are no observations (right/wrong) in the train file, replace 1
or 2 with . (period).
